torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
+        for (size_t i = 0, N = keys.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!keys[i]->type()->isSubtypeOfExt(key_type, &ss)) {
+            throw ErrorReport(key_trees[i])
+                << "Dict keys must contain "
+                << "only a single type, expected: " << key_type->repr_str()
+                << " but found " << keys[i]->type()->repr_str() << " instead.\n"
+                << ss.str();
+          }
+        }
+
+        if (!values.empty()) {
+          // Get the smallest common supertype of all the values. We
+          // need to use the O(N^2) algorithm instead of using
+          // `std::accumulate` to account for the case that two types
+          // cannot themselves be unified, but there exists some mutual
+          // parent type later in `values`. (E.g. T2 <: T1, T3 <: T1,
+          // `values` = {T2, T3, T1}. The real supertype should be `T1`,
+          // not `Any`)
+
+          // List of possible parent types
+          std::vector<TypePtr> seen;
+
+          auto get_greater_type =
+              [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+            if (t1->isSubtypeOf(t2)) {
+              return t2;
+            } else if (t2->isSubtypeOf(t1)) {
+              return t1;
+            } else {
+              return c10::nullopt;
+            }
+          };
+
+          auto value_types =
+              c10::fmap(values, [](Value* v) { return v->type(); });
+
+          for (const TypePtr type : value_types) {
+            bool added = false;
+            for (size_t i = 0; i < seen.size(); ++i) {
+              // Don't bother compare if it's already in our list of
+              // possible supertypes
+              if (type == seen[i]) {
+                continue;
+              }
+              auto maybe_parent = get_greater_type(type, seen[i]);
+              if (maybe_parent && *maybe_parent == type) {
+                seen[i] = type;
+                added = true;
+              }
+            }
+            if (!added) {
+              seen.push_back(type);
             }
           }
-        };
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+
+          TypePtr inferred_type = seen.size() == 1 ? seen[0] : AnyType::get();
+
+          std::stringstream ss;
+          TORCH_CHECK(
+              !type_hint ||
+                  inferred_type->isSubtypeOf(
+                      type_hint->expect<DictType>()->getValueType()),
+              "Type hint for dict values was ",
+              type_hint->repr_str(),
+              " but the greatest common supertype of the given values",
```

Comment:
IIUC, this can only happen when there is at least one value that is not a sub-type of type hint. Is this accurate? If so, maybe we could make the error message simpler and more actionable if we loop through all value types and check their type relationship with type hint. That way, we would be able to tell user which exactly value offends typing rules, instead of erroring on an inferred supertype that user may not have specified themselves.

Reply:
fixed